### DISCUSSION TITLE
Qute Template Rendering with Streaming to Avoid Huge Heap-Memory Consumption in case of Large HTML

### DISCUSSION THREAD

**Discussion by funkrusher** [quarkusio/quarkus#41566] [Q&A] [answered]
Hello Quarkus Team,

i would like to ask, how i could use the Qute Template Engine with a Streaming-Approach,
and if someone can point me in the right direction to do this. 

---
I need this to be able to generate huge HTML-Files that i want to transform to PDF files (containing 100 of pages).
I could not find a streaming approach in the Qute Examples. It seems to me that Qute will render the Template as "String" and i assume that it will load this String completely into Memory during Rendering, which will be a Main-Consumer of Heap-Memory of the Server, and will possibly causing problems with Peaks during Rendering.

I dont know if it is possible to render HTML in Chunks / Streams, but if it is possible, then it should be very low in memory usage because the Stream will be immediately consumed. In the docs there seems to be no streaming support for rendering. See:
- https://quarkus.io/guides/qute-reference#rendering-output

A "OutputStream", "ByteArrayOutputStream" or similar would be great.

Thanks :)

[a comment by an automated account is not shown]

**Comment by mkouba** [comment, marked as answer]
> It seems to me that Qute will render the Template as "String" and i assume that it will load this String completely into Memory during Rendering, which will be a Main-Consumer of Heap-Memory of the Server, and will possibly causing problems with Peaks during Rendering.

As you already found out there are several ways how to trigger rendering and consume the rendered result. For your use case, the `TemplateInstance#consume(Consumer<String>)` method looks like a good match. The consumer is invoked whenever a chunk is available and the returned completion stage that is completed once the rendering finished.

However, keep in mind that in Qute the result tree is always built during rendering. So there would be still some memory overhead involved.

**Comment by funkrusher** [comment]
@mkouba 
thanks, i will try this out, and see if the memory usage goes down with this :)

**Reply by mkouba** [reply]
Let us know if it helps... ;-)

**Reply by funkrusher** [reply]
@mkouba 

i have to admit, i have not tried it out then, because i found a way to work around the issue by splitting the PDF generation in separate parts (flyingsaucer pdf-library allows this), and just make a separate qute template for each such part. 

See my blog-post:
- https://funkrusher.de/en/2024/07/07/pdf-generator-in-java-mit-geringer-heap-memory-auslastung-mit-flyingsaucer-2/
